src/next/useSearchParams.ts

```diff
@@ -1,63 +1,66 @@
-import { useCallback, useEffect, useMemo, useRef } from "react"
-import {
-  usePathname,
-  useSearchParams as useNextSearchParams,
-  useRouter
-} from "next/navigation"
+import { useCallback, useEffect, useRef } from "react"
+import { useSearchParams as useNextSearchParams } from "next/navigation"
 
 type SearchParamsSetterValue =
   | URLSearchParams
   | ((prevSearchParams: URLSearchParams) => URLSearchParams)
 
 type SetSearchParams = (newSearchParams: SearchParamsSetterValue) => void
 
-const useSearchParams = (): [URLSearchParams, SetSearchParams] => {
-  const pathname = usePathname()
-  const router = useRouter()
-  const search = useNextSearchParams()
-
+/**
+ * Returns a function for setting search params in the browser URL.
+ *  - the first call within a single render cycle will add a new entry in the
+ *   history stack via pushState
+ *  - subsequent calls will replace the previous entry in the history stack
+ *    and will receive the previously updated value in setter function
+ *
+ * NOTE: This is intended for use with NextJS, but could work with any framework
+ * that supports direct usage of window.history.pushState / replaceState.
+ */
+const useSetSearchParams = (): SetSearchParams => {
   /**
    * Keep track of whether navigate has been called in the current render cycle
    * to avoid adding extra entries in the history stack.
    */
   const hasNavigatedRef = useRef(false)
-  const searchParams = useMemo(() => new URLSearchParams(search), [search])
-  /**
-   * Keep track of the current searchParams value so that updater functions can
-   * use the current value rather than value from previous render.
-   */
-  const searchParamsRef = useRef(searchParams)
```

Comment:
NextJS keeps `window.location.search` in sync with the router's `useSearchParams`, so replacing the ref with `window.location.search` seems fine, and simplifies the hook.

In general in React,  changes to `window.location` won't trigger a rerender, but it's OK to use inside a callback, as here.